File: bindings/flutter/ios/Classes/HologramViewFactory.swift
import Flutter
import UIKit
import HologramismKit

/// Builds `HologramPlatformView`s for the `hologramism/HologramView` view type,
/// decoding scene/tilt from the StandardMessageCodec creation params.
class HologramViewFactory: NSObject, FlutterPlatformViewFactory {
  private let messenger: FlutterBinaryMessenger

  init(messenger: FlutterBinaryMessenger) {
    self.messenger = messenger
    super.init()
  }

  func createArgsCodec() -> FlutterMessageCodec & NSObjectProtocol {
    FlutterStandardMessageCodec.sharedInstance()
  }

  func create(withFrame frame: CGRect, viewIdentifier viewId: Int64, arguments args: Any?) -> FlutterPlatformView {
    HologramPlatformView(frame: frame, viewId: viewId, args: args, messenger: messenger)
  }
}

/// Wraps the shared `Hologramism.HologramView`. `UiKitView` sets creation params
/// once, so a per-view method channel (`hologramism/HologramView/<id>`) carries
/// later scene/tilt updates when the Dart widget rebuilds.
class HologramPlatformView: NSObject, FlutterPlatformView {
  private let hologram = HologramView()
  private let channel: FlutterMethodChannel

  init(frame: CGRect, viewId: Int64, args: Any?, messenger: FlutterBinaryMessenger) {
    channel = FlutterMethodChannel(
      name: "hologramism/HologramView/\(viewId)", binaryMessenger: messenger)
    super.init()

    hologram.frame = frame
    apply(args)

    channel.setMethodCallHandler { [weak self] call, result in
      guard let self = self else { return }
      switch call.method {
      case "setScene":
        if let dict = call.arguments as? NSDictionary { self.hologram.setScene(dict) }
        result(nil)
      case "setTilt":
        if let dict = call.arguments as? NSDictionary { self.hologram.setTilt(dict) }
        result(nil)
      default:
        result(FlutterMethodNotImplemented)
      }
    }
  }

  private func apply(_ args: Any?) {
    guard let params = args as? [String: Any] else { return }
    if let tilt = params["tilt"] as? NSDictionary { hologram.setTilt(tilt) }
    if let scene = params["scene"] as? NSDictionary { hologram.setScene(scene) }
  }

  func view() -> UIView { hologram }
}

File: bindings/flutter/android/src/main/kotlin/io/github/alexdonh/hologramism/flutter/HologramViewFactory.kt
package io.github.alexdonh.hologramism.flutter

import android.content.Context
import android.view.View
import io.github.alexdonh.hologramism.HologramView
import io.flutter.plugin.common.BinaryMessenger
import io.flutter.plugin.common.MethodChannel
import io.flutter.plugin.common.StandardMessageCodec
import io.flutter.plugin.platform.PlatformView
import io.flutter.plugin.platform.PlatformViewFactory

/**
 * Builds [HologramPlatformView]s for the `hologramism/HologramView` view type,
 * decoding scene/tilt from the StandardMessageCodec creation params.
 */
class HologramViewFactory(
    private val messenger: BinaryMessenger,
) : PlatformViewFactory(StandardMessageCodec.INSTANCE) {

    override fun create(context: Context, viewId: Int, args: Any?): PlatformView =
        HologramPlatformView(context, viewId, args, messenger)
}

/**
 * Wraps the shared [HologramView]. `AndroidView` sets creation params once, so a
 * per-view method channel (`hologramism/HologramView/<id>`) carries later
 * scene/tilt updates when the Dart widget rebuilds.
 */
private class HologramPlatformView(
    context: Context,
    viewId: Int,
    args: Any?,
    messenger: BinaryMessenger,
) : PlatformView {

    private val hologram = HologramView(context)
    private val channel = MethodChannel(messenger, "hologramism/HologramView/$viewId")

    init {
        apply(args)
        channel.setMethodCallHandler { call, result ->
            when (call.method) {
                "setScene" -> {
                    (call.arguments as? Map<String, Any?>)?.let { hologram.setScene(it) }
                    result.success(null)
                }
                "setTilt" -> {
                    (call.arguments as? Map<String, Any?>)?.let { hologram.setTilt(it) }
                    result.success(null)
                }
                else -> result.notImplemented()
            }
        }
    }

    @Suppress("UNCHECKED_CAST")
    private fun apply(args: Any?) {
        val params = args as? Map<String, Any?> ?: return
        (params["tilt"] as? Map<String, Any?>)?.let { hologram.setTilt(it) }
        (params["scene"] as? Map<String, Any?>)?.let { hologram.setScene(it) }
    }

    override fun getView(): View = hologram

    override fun dispose() {}
}
